File: api/sync/src/main/java/com/coditory/sherlock/SherlockWithConnector.java
package com.coditory.sherlock;

import com.coditory.sherlock.DelegatingDistributedLock.AcquireAction;
import com.coditory.sherlock.DelegatingDistributedLock.ReleaseAction;
import com.coditory.sherlock.DistributedLockBuilder.LockCreator;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import static com.coditory.sherlock.Preconditions.expectNonNull;

final class SherlockWithConnector implements Sherlock {
    private final Logger logger = LoggerFactory.getLogger(SherlockWithConnector.class);
    private final DistributedLockConnector connector;
    private final LockDuration defaultDuration;
    private final OwnerIdPolicy defaultOwnerIdPolicy;

    SherlockWithConnector(
            DistributedLockConnector connector,
            OwnerIdPolicy defaultOwnerIdPolicy,
            LockDuration defaultDuration) {
        this.connector = expectNonNull(connector, "Expected non null connector");
        this.defaultOwnerIdPolicy =
                expectNonNull(defaultOwnerIdPolicy, "Expected non null defaultOwnerIdPolicy");
        this.defaultDuration =
                expectNonNull(defaultDuration, "Expected non null defaultDuration");
    }

    @Override
    public void initialize() {
        logger.debug("Initializing sherlock infrastructure");
        connector.initialize();
    }

    @Override
    public DistributedLockBuilder<DistributedLock> createLock() {
        return createLockBuilder(connector::acquire, connector::release);
    }

    @Override
    public DistributedLockBuilder<DistributedLock> createReentrantLock() {
        return createLockBuilder(connector::acquireOrProlong, connector::release);
    }

    @Override
    public DistributedLockBuilder<DistributedLock> createOverridingLock() {
        return createLockBuilder(connector::forceAcquire, (id, __) -> connector.forceRelease(id));
    }

    @Override
    public boolean forceReleaseAllLocks() {
        return connector.forceReleaseAll();
    }

    private DistributedLockBuilder<DistributedLock> createLockBuilder(
            AcquireAction acquireAction,
            ReleaseAction releaseAction) {
        return new DistributedLockBuilder<>(createLock(acquireAction, releaseAction))
                .withLockDuration(defaultDuration)
                .withOwnerIdPolicy(defaultOwnerIdPolicy);
    }

    private LockCreator<DistributedLock> createLock(
            AcquireAction acquireAction,
            ReleaseAction releaseAction) {
        return (lockId, duration, ownerId) ->
                createLockAndLog(acquireAction, releaseAction, lockId, ownerId, duration);
    }

    private DistributedLock createLockAndLog(
            AcquireAction acquireAction,
            ReleaseAction releaseAction,
            LockId lockId,
            OwnerId ownerId,
            LockDuration duration) {
        DistributedLock lock = new DelegatingDistributedLock(
                acquireAction, releaseAction, lockId, ownerId, duration);
        logger.debug("Created lock: {}", lock);
        return lock;
    }
}

File: api/coroutine-connector/src/main/kotlin/com/coditory/sherlock/coroutines/SherlockWithConnector.kt
package com.coditory.sherlock.coroutines

import com.coditory.sherlock.DistributedLockBuilder
import com.coditory.sherlock.DistributedLockBuilder.LockCreator
import com.coditory.sherlock.LockDuration
import com.coditory.sherlock.LockId
import com.coditory.sherlock.OwnerId
import com.coditory.sherlock.OwnerIdPolicy
import com.coditory.sherlock.Preconditions
import com.coditory.sherlock.coroutines.DelegatingDistributedLock.AcquireAction
import com.coditory.sherlock.coroutines.DelegatingDistributedLock.ReleaseAction
import org.slf4j.LoggerFactory

class SherlockWithConnector(
    connector: SuspendingDistributedLockConnector,
    defaultOwnerIdPolicy: OwnerIdPolicy,
    defaultDuration: LockDuration,
) : Sherlock {
    private val logger = LoggerFactory.getLogger(SherlockWithConnector::class.java)
    private val connector: SuspendingDistributedLockConnector
    private val defaultDuration: LockDuration
    private val defaultOwnerIdPolicy: OwnerIdPolicy

    init {
        Preconditions.expectNonNull(connector, "connector")
        Preconditions.expectNonNull(defaultOwnerIdPolicy, "defaultOwnerIdPolicy")
        Preconditions.expectNonNull(defaultDuration, "defaultDuration")
        this.connector = connector
        this.defaultOwnerIdPolicy = defaultOwnerIdPolicy
        this.defaultDuration = defaultDuration
    }

    override suspend fun initialize() {
        logger.debug("Initializing sherlock infrastructure")
        connector.initialize()
    }

    override fun createLock(): DistributedLockBuilder<DistributedLock> {
        return createLockBuilder({ connector.acquire(it) }) { lockId: LockId, ownerId: OwnerId ->
            connector.release(lockId, ownerId)
        }
    }

    override fun createReentrantLock(): DistributedLockBuilder<DistributedLock> {
        return createLockBuilder({ connector.acquireOrProlong(it) }) { lockId: LockId, ownerId: OwnerId ->
            connector.release(lockId, ownerId)
        }
    }

    override fun createOverridingLock(): DistributedLockBuilder<DistributedLock> {
        return createLockBuilder({ connector.forceAcquire(it) }) { id: LockId, _: OwnerId ->
            connector.forceRelease(id)
        }
    }

    override suspend fun forceReleaseAllLocks(): Boolean {
        return connector.forceReleaseAll()
    }

    private fun createLockBuilder(
        acquireAction: AcquireAction,
        releaseAction: ReleaseAction,
    ): DistributedLockBuilder<DistributedLock> {
        return DistributedLockBuilder(createLock(acquireAction, releaseAction))
            .withLockDuration(defaultDuration.value)
            .withOwnerIdPolicy(defaultOwnerIdPolicy)
    }

    private fun createLock(
        acquireAction: AcquireAction,
        releaseAction: ReleaseAction,
    ): LockCreator<DistributedLock> {
        return LockCreator<DistributedLock> { lockId: LockId, duration: LockDuration, ownerId: OwnerId ->
            createLockAndLog(acquireAction, releaseAction, lockId, ownerId, duration)
        }
    }

    private fun createLockAndLog(
        acquireAction: AcquireAction,
        releaseAction: ReleaseAction,
        lockId: LockId,
        ownerId: OwnerId,
        duration: LockDuration,
    ): DistributedLock {
        val lock: DistributedLock =
            DelegatingDistributedLock(acquireAction, releaseAction, lockId, ownerId, duration)
        logger.debug("Created lock: {}", lock)
        return lock
    }
}
